doc/client_channel_specification.md

```diff
@@ -0,0 +1,809 @@
+-- i.e., policies like [round\_robin](#round_robin) that balance across
+endpoints will delegate to a pick\_first child policy to choose an
+address within each endpoint.
+
+#### round\_robin
+
+The round\_robin policy spreads load across a set of endpoints by
+simply sending one RPC to each endpoint in order, wrapping around to
+the beginning of the list when it hits the end.
+
+For each endpoint, the round\_robin policy creates a
+[pick\_first](#pick_first) child policy, which will connect to one of
+the addresses for that endpoint.  When the round\_robin picker wants to
+send an RPC to a particular endpoint, it delegates to the picker from
+the pick\_first child policy for that endpoint.
+
+The policy attempts to maintain a connection to every endpoint at all
+times.  When it receives the initial resolver update, it will create
+the pick\_first child policy for each endpoint, and the pick\_first
+child will immediately start to connect.  If at any point a pick\_first
+child reports IDLE state, round\_robin will immediately call the child's
+ExitIdle() method, thus triggering it to attempt to reconnect.
+
+The round\_robin policy determines what connectivity state to report
+to its parent by aggregating the connectivity states of the endpoints
+(i.e., the states reported by the pick\_first child policies),
+using the normal aggregation rules described in [Channel Connectivity
+States](#aggregated-connectivity-states) below.  However, note that for
+the purposes of this aggregation, endpoints reporting IDLE are treated as
+if they had reported CONNECTING, because round\_robin always immediately
+triggers a connection attempt on the child in that state anyway.
+Specifically (first match wins):
+
+* If there is at least one child in state READY, report READY.  In this case,
+  the picker will round-robin over all subchannels in READY state.  Note that
+  whenever a new picker is returned, it will start from a random index in the
+  list of endpoints.  The picker will need appropriate synchronization to
+  maintain the index as it processes picks.
+* If there is at least one child in state CONNECTING (or IDLE, as mentioned
+  above), report CONNECTING.  In this case, the picker will queue picks.
+* Otherwise, all subchannels are in TRANSIENT\_FAILURE state, so report
+  TRANSIENT\_FAILURE state.  In this case, the picker will fail all picks
+  with an RPC status that includes the error from the most recent subchannel
+  connection failure.
+
+If round\_robin receives a resolver update with an empty address list,
+it will report TRANSIENT\_FAILURE and return an error telling its parent
+that it rejected the update.  The channel will pass this error back to
+the resolver with the health status callback.  For polling resolvers,
+this will cause the resolver to re-resolve with appropriate backoff.
+
+Whenever an endpoint reports TRANSIENT\_FAILURE or IDLE, round\_robin
+requests re-resolution via the helper.
```

Comment:
This happens inside the child PF policy only, though, right?

Reply:
Yup.  Clarified.